cluster-api/pkg/controller/machineset/reconcile_test.go

```diff
@@ -86,11 +91,45 @@ func TestMachineSetControllerReconcileHandler(t *testing.T) {
 		{
 			name:                "scenario 6: the current machine has different labels than the given machineSet, thus a machine is created.",
 			startingMachineSets: []*v1alpha1.MachineSet{createMachineSet(1, "foo", "bar2", "acme")},
-			startingMachines:    []*v1alpha1.Machine{machineWithDifferentLabels(createMachineSet(1, "foo", "bar1", "acme"), "bar1")},
+			startingMachines:    []*v1alpha1.Machine{setDifferentLabels(machineFromMachineSet(createMachineSet(1, "foo", "bar1", "acme"), "bar1"))},
 			machineSetToSync:    "foo",
 			namespaceToSync:     "acme",
 			expectedActions:     []string{"create"},
 		},
+		{
+			name:                "scenario 7: the current machine is missing owner refs, machine should be adopted.",
+			startingMachineSets: []*v1alpha1.MachineSet{createMachineSet(1, "foo", "bar2", "acme")},
+			startingMachines:    []*v1alpha1.Machine{machineWithoutOwnerRefs(createMachineSet(1, "foo", "bar1", "acme"), "bar1")},
+			machineSetToSync:    "foo",
+			namespaceToSync:     "acme",
+			expectedActions:     []string{"update"},
+			expectedMachine:     machineFromMachineSet(createMachineSet(1, "foo", "bar2", "acme"), "bar1"),
+		},
+		{
+			name:                "scenario 8: the current machine has different controller ref, do nothing.",
```

Comment:
I am getting kinda confused on expected behavior. I understand doing nothing if you control the machine and creating if you do not own the machine. However, two cases of not owning the machine is doing different things here. Scenarios 8 & 6 seem to be both about not owning the machine yet one results in a create and one does nothing. Am I missing something here?

Reply:
In scenario 6, the selector labels don't match, the machine is not passed to the reconcile function, so the reconcile function is passed an empty machine list, it creates a new machine.
In scenario 8, the machine is owned by a different machineset. This is a trickery scenario, as this implies two machinesets have been created using the same selector label. To be honest, I am not sure what behavior is correct here. thoughts?

Reply:
@k4leung4 @jessicaochen  You should never overlap selectors. The quote bellow is from [Deployment documentation](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#creating-a-deployment)

> Note: You must specify an appropriate selector and Pod template labels in a Deployment (in this case, app: nginx). Do not overlap labels or selectors with other controllers (including other Deployments and StatefulSets). Kubernetes doesn’t stop you from overlapping, and if multiple controllers have overlapping selectors those controllers might conflict and behave unexpectedly.

The ReplicaSet [behavior](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-selector) is further explained by this quote:

> Also you should not normally create any pods whose labels match this selector, either directly, with another ReplicaSet, or with another controller such as a Deployment. If you do so, the ReplicaSet thinks that it created the other pods. Kubernetes does not stop you from doing this.
If you do end up with multiple controllers that have overlapping selectors, you will have to manage the deletion yourself.

Reply:
@mvladev - Thanks for the context!

@k4leung4 - I think you should either adopt the machine or treat it like you do not own it (and create a replacement) instead of this in-between do nothing approach. The prior makes things easier for cluster bootstrap. The latter would cause less back and forth fighting if there are two machine sets with the same label.

@krousey - Do you have any thoughts on the right approach for this case?

Reply:
@jessicaochen I have no problem with the current scenarios. Scenario 6, the machine set creates a new machine because none exist that match it's labels.

Scenario 8, it refuses to fight with another machine set that already has an owner ref on it. Scenario 8 is either a bad config or perhaps an old MachineSet that was foreground deleted (allow deletion of the parent before deleting the children) and the GC hasn't gotten around to deleting this particular machine yet. Either way, doing nothing is probably the right move.